### PR TITLE
Paper: Geodynamic simulations in HPC with Python (Creati)

### PR DESCRIPTION
**Pull request by ncreati**
Nicola Creati paper 

[Edit by jaberg] http://nipy.bic.berkeley.edu:5000/download/26

### CODE REVIEW

**Comment by ashtonmsu** [comment]
This review follows the suggested review criteria (https://github.com/scipy-conference/scipy_proceedings/blob/master/review_criteria.md)

1. Written Quality
The paper is written in mostly sound technical English. The authors avoid jargon, and the paper is clearly organized. Some fixes are indicated below:

Typos:
"The porting of PyGmod form the commodity PC to a supercomputer" - should be "from"
"nodes of the grid and viceversa" - "vice versa", though simply "and back" might be clearer for some readers.
"# Pysical constants" (in code (really config file) snippet) "Physical"
"Since the visualization of markers as a cloud of points can be critical when they are more than one million,"  - Since the visualization of over a million markers as a cloud of points can be challenging,"
"in for-loops ad the atomic nature" - "as"
"not suited for computational intensive purpose and" - "purposes"

2. Technical Content
I am not a geoscientist, nor do I work on HPC problems. However, the scientific and technical context of the work is clearly articulated and logically consistent. 

3. Novelty & Impact
The contribution is the porting of code to conduct geosynamic simulation in parallel to Python. As with most complex projects these days, the contribution depends on many existing packages, which the authors acknowledge.

4. Verifiability
The paper is well-referenced, with the authors' work situated within existing research.
The paper demonstrates the authors' contributions with representative code chunks and output.

5. Other
I think all is fine here.

6. Length
Fine.

7. Figures and Code Snippets
I am not sure why the figures are presented on page 5, sandwiched between some references and the remainder of the references. Perhaps that could be situated inline or otherwise better organized?
Code snippets are clear and readable.

**Comment by aashish24** [comment]
Just wanted to inform that I have started the review and post my notes here soon.

**Comment by cbcunc** [comment]
# Written Quality
## Abstract
* Consider dropping phrase at end of sentence: "that can benefit  of  it." Language is stilted and redundant with previous adjective "some" in same sentence. Use "benefit from it" if you feel you must keep it.

## Introduction
* Change  "since" in "geodynamics is necessary since tectonic processes are too slow" to "because." "Since" refers to a point in time whereas "because" refers to causality.
* Change "had been developing" in "In  the  last  30  years  numerical  geodynamic  modeling  has
been  developing  very  rapidly" to "has developed." "In the last 30 years" refers to a past time period rather than on ongoing time period.
* Comma before "such as" in " known  problems  in  computational science  such  as." "Such as" is followed by a *non-restrictive* clause. See Chicago Manual of Style.
* Comma after "and  other  thermodynamic  variables" to end non-restrictive clause.
* Change "rocks rheology" to "rock rheology." In this context, rock is an adjective rather than refering to some collection of rocks.
* Change "architecture" to "architectures" in "computer clusters of different architecture." There is more than one architecture in the collection of clusters.
* Comma after "however" in "However to take full advantage" at beginning of sentence.
* Run-on sentence: "However to take full advantage of the power of these machines parallel algorithms and software packages must be developed and often geoscience researchers do not have enough knowledge  to  build  and  debug  parallel  software." Change to "To take full advantage of the power of these machines, parallel algorithms and software packages must be developed. However, geoscience researchers often do not have enough knowledge  to  build  and  debug  parallel  software."
* Comma between clause and predicate: "While  in  the  last fifteen  years  several  numerical  methods  and  libraries  have been  developed  to  solve  many  of  the  equations  needed  to model geodynamic problems**,** almost all of them are written in C/C++  or  Fortran.
* "Sometimes  geodynamic  modeling  is  done with commercial software [Pas02], [Jar11] to reduce the effort of solving  ~~the~~  equations,  **as well we as** writ**ing**  and  debug**ing**  a  new  application." That is, remove the, distinguish verbs with different objects, and bring verbs into tense agreement.
* Run-on sentence confusing two ideas: "A numerical geodynamic modeling application, PyGmod, has been developed in the last twelve months in the context of the PRACE-OGS research project about the development of HPC applications for oceanographic and geophysical numerical simulations, by exploiting several years of Python experience in remote sensing and geoscience topics." Break into: "By exploiting several years of Python experience **with** remote sensing and geoscience topics, a numerical geodynamic modeling application, PyGmod, has been developed in the last twelve months. **The development  of PyGmod occurred within** the PRACE-OGS research project, **which is concerned with** ~~the development of~~ HPC applications for oceanographic and geophysical numerical simulations"

## Scientific Context
* Change "rifts" to "rift" in "poor understanding of rifts initiation." In this context, rift is an adjective rather than referring to some collection of rifts.
* "Understanding how, why and when  rifts  initiate  and  localize  is  important  ~~to~~ **for defining**  ~~every~~ factor  controlling  their  dynamics."
* "The relative  importance  of these factors during the inception and earliest development of a new rift  **is** controversial." Subject/verb agreement. The subject is "importance." Importance **is** controversial. Alternatively, you may mean plural importances, in which case the relative importances of these factors are controversial, as each factor may have its own importance which is relative to the other importances.

## Core Development
* The first Python version was  serial,  *was* based  on  the  Numpy  [Van11]  and  Scipy  [Jon01] packages  to  manage  arrays,  **and solved**  the  governing  differential equations ~~and to do all the math needed~~." Verbs in a list need to agree on tense. If the equations were solved, then all the math was done.
* "The  porting  of PyGmod  form  the  commodity  PC  to  a  supercomputer  was not straightforward ~~since~~ **because** every supercomputer has a different hardware  architecture."
* "The  MPI  message-passing  system  has been  adopted  while  the  Python  multiprocessing  module  was avoided because it is not available on all supercomputer**s** (e.g. the  IBM  Blue  Gene  series  [Gil13])." Pluralize supercomputer(s) when preceded by "all."
* "The  management  of parallel  distributed  arrays  has  been  possible  thanks  to  the Global Array Toolkit (GA) library [Nie06] ~~that~~ which is available in Python as ga4py." You aren't distinguishing the GA library from another one *that* is available in another package, so it is the GA *which* is available in the ga4py package (non-restrictive which, sentence makes sense without the clause).

## PyGmod Structure
* "The  deformation  of  rocks  is  modeled  as  flow  of variable  viscous  material  as  in  computational  fluid  **dynamics** [Ver07]." Pluralize dynamics.
* "The governing equations reflect the conservation laws of physics:" Use commas after each of them items the following list instead of semicolons and terminate the last item with a period as it is the end of the sentence incorporating the list.
* "These equations are coupled with rocks rheological laws..." Use the singular "rock" in this context.
* "...moving-marker Lagrangian approach..." Hyphenate "moving-marker" as it is a compound adjective to the adjective "Lagrangian."
* "...while the markers are a cloud of points ~~that~~ **which** covers the grid." Clause is non-restrictive (sentence makes sense without it). and so use "which."
* "The markers in cell methodology ~~is~~ **are**  characterized..." Subject/verb agreement. The makers *are* characterized.
* "..to the nodes of the grid and viceversa..." "Vice versa" are two Latin words, not one. Separate by a space.
* "For example the IBM BG/Q at CINECA..." Comma after "For example."
* "GA  greatly  helps  to  create,  distribute  and  manage  all  the arrays,  both  1D  and  2D  providing  a  shared  memory  style programming environment in the context of a distributed array data  structures." Comma after "2D." Hyphenate "memory-style" compound adjective.
* "the  remote process(es) ~~that~~ **which** own(s) the referenced data." Non-restrictive optional clause.
* "PyGmod  uses  a  modified  GA  version  ~~that~~  **which** implements  the ARMCI-MPI  [Armci]  RMA  (Remote  Memory  Access)  **one-sided**  communication  ~~since~~ because  the  standard  GA  implementation, available at the time of the development of PyGmod, worked only on few hardware architectures."
* "...split the 2D domain in **Cartesian** blocks..." Capitalize Cartesian.
* "...assign  to  each  block  the  markers ~~that~~ **which**  belong  to  it."
* "...markers inside each Cartesian block..." Capitalize Cartesian.
* " In  each  time  iteration,  most  of  the calculation  is  done..." Comma after "iteration."
* "PyGmod uses direct equation solvers to achieve accurate solutions like MUMPS [Ame00] or Superlu [Li03]  ~~since~~ because  the  problem  is  2D  and  current  supercomputers provide enough memory."
* "...PETSc  sparse  arrays  and  vectors  must  be  extracted  and  the corresponding  data  block  must  be  retrieved  as  Numpy  **arrays**..." Pluralize "arrays."
* "~~Here~~ **In this example,** *L* is a PETSc distributed bi-dimensional array, *dofs* is the degree  of  freedom  of  the  system,  **and** *ynum*  is  the  total  number of rows of the model." Italicize identifier names, Hyphenate "bi-dimensional." Separate the last item in the list with "and."
* "The local array ~~is then~~ used to fill the local PETSc portion of the sparse arrays and vectors. The local solution of the equations, a Numpy array, is then put back in the corresponding global array." Use "is then" only once at the conclusion of a sequence as the phrase is final.
* "The *MPILog* class uses ~~the~~ *MPI.File~ **objects** ~~and methods~~ to write logging information to standard output..." Italicize identifiers. If the class creates or uses more the one instance of MPI.File, the refer to those objects in plural without further reference to the multople instances' methods. Otherwise, if referring to the methods of a single MPI.file instance you may say, " The *MPILog* class uses *an* *MPI.File~ **object's**methods," where "object's" is possessive.
* "Six  log  levels  are  defined:  *critical*,  *error*,  *warning*, *notice*, *info* and *debug*, ~~being~~ **with** *warning* **as** the default". Edited for sense. It would be nice to italicize the warning levels so as to emphasize their descriptive nature. Consider dropping the sentence altogether as these are standard log levels and are not descriptive of PyGmod in particular.
* "Each MPI process  writes  its  log  ~~on~~ **to**  the  same  file  in  a  different  ~~colors~~ **color**." Color is singular for each process.
* "Log level as well as processor ID number can be filtered out." This needs to be in a separate sentence instead of compounded to the previous predicate.
* "Log calls are ~~inserted~~ **invoked** according to the following syntax:"
* "~~Here~~ **In this example,** *inmsg* is the message string, *watch* is ~a~ **the** list of processors ~~that we are interested in~~ **to which the message applies**, *rank* is a switch to hide the processor rank  from  the  message,  *mono*  disables  colorized  messages ~~reverting to monochrome~~**,** and *level* defines the minimum level at which the message will be printed." Italicize parameter names. Comma before last item in list. Delete redundant phrase. Other edits as noted.
* "... boundary conditions,  etc.." "Etc." is an abbeviation requiring its own period. The end of the sentence requires another period.
* "Units  of  measurements  can  be  included  in the  configuration  file  ~~since~~ because  the  parsing  system  implemented ~~is  able  to  convert~~ converts  the  units  to  the  right  ones  needed  by PyGmod  ~~while~~ checking  also  for  dimensionality  consistency." Unnecessary passive voice. Missing preposition.
* "Each time iteration is  saved  in  a  different  hdf5  file  (evolution  step)  to  avoid  ~~too~~ large files." "Too large" for what?  You may "...to avoid creating files too large fo the filesystem," if that applies.
* "~~There is also~~ **A** main output file **also** ~~containing~~ **contains** a copy of  the  configuration  ~~file ~~ ~~used  to~~ **which** ~~generate~~ **generated**  the  simulation  ~~that keeps track of~~ **for** the ~~whole~~ **entire** evolution." Eschew passive voice. For clarity, avoid sequences of restrictive clauses. An evolution is not a whole, but is a sequence of steps which collectively constitutes an entirety.
* "RTV code is based on Matplotlib [Hun07] and ~~draws images~~ **plots** ~~of~~ some of
the quantities calculated in the simulation (Figure 2)." If an image was drawn of a quantity, I might expect to see a quantity like "5" rendered as a image instead of text.
* "~~Since~~ Because the visualization  of  markers  as  a  cloud  of  points  can  be  critical when  ~~they  are~~ **the points number**  more  than  one  million..." Alternately say, "when **there** are more than one million **points**." And be sure to say more than one million *what* (points or markers).
* "Thus, each processor..." Comma following "Thus."
* "The  interpolation  uses  the  *griddata*  module  of  Scipy  with  a nearest  neighbors  switch" It would be nice to see module and package names italicized (or even botd italicized) through the document as these are proper names like titles.
* "RTV  can  plot  data  from  a  real- time  simulation  showing  the  current  evolution  step  or  ~~from~~ historical data."

## Performance
* "PyGmod ~~has been~~ was built using optimized third party libraries..." Be emphatic. "Has been" implies that maybe it is not always built that way, or is possibly no longer built that way.
* "...the size of the problems used to develop the code  was  not  so  big **so as** to  require  further  optimization." Infinitive without "in order" or "so as."
* "Tests  proved  that  ~~markers~~ **marker**  interpolation is  a  critical  operation  that  can  take  a  large  amount  of  time." Singular adjective.
* "Interpolation is done in for-loops ~~ad~~ **and** the atomic nature of the algorithm  used  forbids  the  use  of  Numpy  methods." Spelling/typo.
* "they tend to be ~~of~~ on the order of millions..."
* "...and every time iteration step calls ~~several times~~ the interpolation methods **several times**." Awkward subject/object split by adverbial phrase.
* "~~As~~ Because Python loops are inherently  slow,  Cython  has  been  used  to  speed  up  markers interpolation. " The speed of Python loops is not a metaphor for using Cython. It is a reason for using Cython.
* "...with minor modifications,..." You list more than one modification, so make "modifications" plural.
* "The net increase of speed with this  simple  ~~operation~~ **technique**  is  almost  three  orders  of  magnitude (Table 1)." A "operation" is either an ongoing process or a calculation.
* "The  performance  has  been  further  improved  by activating  ~~threading~~ **threads**  in the interpolation methods." Or even better, "The  performance  has  been  further  improved  by threading ~~in~~ the interpolation methods." But please don't use a double gerund. If you were instead referring to the *threading* standard library module, distinguish it so with italics or bold italics.
* "Thanks  to Cython  [Beh11]**,**..." Comma after Cython reference.
* " Loops  are  split  ~~in~~ **into** threads..." If i understand correctly, the threading occurs by splitting the loops into concurrent threads with each thread handling only a portion of the entire virtual loop's iterative domain, rather than splitting up the loop inside a thread.
* "...and further optimize ~~further~~ the code." Awkward split of subject/object by adverb.

## Final Remarks
* "Available third part Python packages greatly ~~reduce~~ reduced the work needed to parallelize the algorithms."
* "...and ~~take charge~~ **handle** of all the **necessary** communication ~~needed~~."
* "Pure  Python  code  **absolutely**  can  ~~absolutely~~  be ~~further~~ optimized **further by** using different switches or methods provided by  ~~the~~  external  packages  (e.g.  equation  solvers)." Don't split the infinitives. Consider deleting the adverb "absolutely" because the you can't be absolute about further optimizations you haven't performed yet.
* "PyGmod  is ~~a~~  young  code  that  works  without  any ~~line  of~~  C  or  Fortran."
* "It  can  be  modified  with  minor  effort,  adapted  to  the  needs of  the  research**,**  and  extended  including  other  geodynamic phenomena  like  melting,  fluid  migration,  phase  changes,  etc.**.**" Comma after "research" (before last item in list). Period at end of sentence after period ending "etc." abbreviation.
* "Open-source  **and** efficient  libraries  and  packages  available  in  the Python  universe  ~~allow  to~~  overcome  the  myth  that  Python is  only  a  scripting  language  not  suited  for  ~~computational~~ **computationally**  intensive purpose ~~and~~ **or** that cannot be used on HPC facilities."

## Table 1
* "**Performance** comparision between interpolation code ~~speedup~~ adopting Cython and threading." "Speedup" is not a word unless hyphenated. Better to use "performnace" as an adjective for "comparison" in this case.

## Figure 1
* "Example of **on-screen** log output ~~on screen of~~ by four MPI tasks run**ning** with log level "info"." It's not on the screen of four tasks. It's output on the screen.

# Technical Content
* The computational content of the paper cannot be verified as the code location is not cited. Closed code is not scientific code.
* Several introductory statements about the history of geodynamic computation have no citations.

# Novelty or Impact
* The code appears to advance a scientific domain, but with well known methods.
* Novelty may be found in the combination of packages integrated.

# Verifiability
* No data source is specified.
* No workflow is specified.

# Other Requirements
* Except as previously noted for code, data, and workflow verifiability, no objections.

# Length
* Good!

#  Figures
* Inter-mized with references. Makes references hard to follow.

# Code snippets
* Overflows  columns
* Fails PEP 8 (expressions without readable spacing)

**Comment by aashish24** [comment]
1. Written Quality:
This work mostly did a nice job describing the problem. The paper is nicely organized. Some aspect of writing could be improved though.
 
Introduction
The sentence below is too long. 
A numerical geodynamic modeling application, PyGmod, has been developed in the last twelve months in the context of the PRACE-OGS research project about the development of HPC applications for oceanographic and geophysical numerical simulations, by exploiting several years of Python experience in remote sensing and geoscience topics.
Final Remarks
“Available third part Python packages” should be “Available third party Python packages” 
Overall
There are some other minor corrections needs to be made. I suggest author review the entire paper again and fix these minor language issues (commas, typos etc). Also, looks like PyGMod structure is a very large section. It could possibly use sub-sections.
2. Technical Content
Technical content has good detail with references. However, it would have been nice to show strong and weak scaling performance of the system. For a reference please see this article: https://www.nersc.gov/assets/NUG-Meetings/IntroMPIApplicationScaling.pdf 
Novelty & Impact
The work presented is novel in a sense that it was able to combine various python pacakges and Cython successfully to enable parallelism for Geodynamic simulations. This paper has lots of detail. 
Verifiability
I couldn’t find the link to the data or the source. 
Other

Length
Length is fine for this paper. Although, it could very well be a long paper as well. 

Figures and Code Snippets

**Comment by aashish24** [comment]
Another important point is that since this work has used a lot of packages, it would have been nice to see a stack diagram or architecture diagram to depict relationship between different components.

**Comment by ncreati** [comment]
We thank very much all the reviewers for very constructive comments which helped to improve the quality of the paper.
We rearranged the paper following the reviewers suggestions as much as possible.

A layer diagram of our application has been added to show the third party libraries we used. We do not include any flow chart of the 2D thermo-mechanical code because it will need a long, technical, and detailed explanation.

More references have been added about plate tectonics in the introduction.

Code snippets have been reviewed following PEP8.

We state in the paper that we are currently running tests about scalability on a supercomputer (GALILEO: http://www.hpc.cineca.it/hardware/galileo ) at CINECA. We have no results to show yet.

PyGmod is still in testing phase (it cannot be used for production) and we did not decide yet if the code will be opensource. At present the code consists of over 10k lines.

About data availability: PyGmod creates synthetic geodynamic simulations from an input configuration text file. Output is a collection of hundreds hdf5 files whose structure is still under development.

**Comment by katyhuff** [comment]
@aashish24 @cbcunc @ashtonmsu If you have a chance to confirm whether the changes by the authors satisfy your comments, it would be greatly appreciated if you could let myself and my co-chair know of your approval with a message like "@katyhuff @jaberg , this paper is ready" or otherwise.

Thanks!

**Comment by shjb** [comment]
Looks like comments have been addressed, thanks everyone for great work!

**Comment by jaberg** [comment]
Looks like comments have been addressed, good job everyone!

**Comment by aashish24** [comment]
@katyhuff this paper is ready! Good stuff. Many thanks for taking care of reviews.
